designs/simultaneous-edit/README.md

```diff
@@ -0,0 +1,121 @@
+---
+state: draft
+---
+
+# Simultaneous Flow Editing
+
+## Summary
+
+With current Node-RED, multiple clients can connect to the same runtime and can edit the same flow independently.  However, editing results by multiple clients are not synchronized, so conflicts of flows occur during deployment of the flow.  There is a feature to support merging flow conflicts referring node property values, but it is difficult for end users to merge node properties in proper way.
+
+![conflict-detection.png](/Users/nisiyama/src/Curry/Git/designs/designs/simultaneous-edit/images/conflict-detection.png)
+
+![review-change.png](/Users/nisiyama/src/Curry/Git/designs/designs/simultaneous-edit/images/review-change.png)
+
+Therefore, this design document proposes a new Node-RED editor feature that supports multiple clients to edit the same flow simultaneously.
+
+## Authors
+
+- Hiroyasu Nishiyama
+
+## Proposal
+
+Here, we consider the following two methods as candidates for simultaneous editing support of a Node-RED flow:
+
+1. Synchronize editing actions among clients,
+
+2. Allow locking a tab by a client to make single client can only edit the tab.
+
+### Method1: Edit Action Synchronization
+
+#### 1-a: Requirements
+
+- The flow edit result of a client is immediately reflected to the flow edited on workspaces of other clients.
+
+#### 1-b: Issues
+
+- How to reflect flow editing state among clients?
+
+- How to handle other kind of state changes of Node-RED including editor and runtime (e.g. node installation, execution result, etc.)?
+
+#### 1-c: Implementation
+
+- Extends the edit history information currently used for *undo* / *redo* of flow editing operations, and sends editing action information to other clients via the runtime.
+
+- The client that received the edit action reproduces the flow edit result by executing the received edit action on the client browser.
+  
+  - In order to this properly work, ordering of edit actions are required to guarantee the order in which the edit operations are applied.
+  
+  - Serialization of actions are provided as a feature of the runtime.
+    
+    Though, there are performance concerns about this with large number of clients or slow network.
+  
+  - Changes other than the flow should also be treated as editing actions to be sent.  However, it is necessary not to execute anything that involves a run-time operation.
+
+![synchronize-actions.png](/Users/nisiyama/src/Curry/Git/designs/designs/simultaneous-edit/images/synchronize-actions.png)
+
+- We may need to synchronize client state at some point by sending refresh request to all client.
+
+- Whether to use sync mode can be specified in `settings.js`.
```

Comment:
Yes - this would have to be an opt-in feature.

I also wonder whether this should be provided by a plugin rather than part of the core. It would require some additional APIs to be exposed in core for sure, but it may be a cleaner approach - particular for those who embed Node-RED and don't want this type of functionality at all.

Reply:
Yes. Various event extensions will be required in Node-RED side, but I think it can be implemented as a plugin.